pow/ethash.go

```diff
@@ -0,0 +1,579 @@
+		file, err = os.OpenFile(path, os.O_RDONLY, os.ModePerm)
+	}
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	if write {
+		mem, err = mmap.Map(file, mmap.RDWR, 0)
+	} else {
+		mem, err = mmap.Map(file, mmap.RDONLY, 0)
+	}
+	if err != nil {
+		file.Close()
+		return nil, nil, nil, err
+	}
+	// Yay, we managed to memory map the file, here be dragons
+	header := *(*reflect.SliceHeader)(unsafe.Pointer(&mem))
+	header.Len /= 4
+	header.Cap /= 4
+
+	return file, mem, *(*[]uint32)(unsafe.Pointer(&header)), nil
+}
+
+// cache wraps an ethash cache with some metadata to allow easier concurrent use.
+type cache struct {
+	epoch uint64 // Epoch for which this cache is relevant
+
+	dump *os.File  // File descriptor of the memory mapped cache
+	mmap mmap.MMap // Memory map itself to unmap before releasing
+
+	cache []uint32   // The actual cache data content (may be memory mapped)
+	used  time.Time  // Timestamp of the last use for smarter eviction
+	once  sync.Once  // Ensures the cache is generated only once
+	lock  sync.Mutex // Ensures thread safety for updating the usage time
+}
+
+// generate ensures that the cache content is generated before use.
+func (c *cache) generate(dir string, limit int, test bool) {
+	c.once.Do(func() {
+		// If we have a testing cache, generate and return
+		if test {
+			c.cache = make([]uint32, 1024/4)
+			generateCache(c.cache, seedHash(c.epoch*epochLength+1))
+			return
+		}
+		// If we don't store anything on disk, generate and return
+		size := cacheSize(c.epoch*epochLength + 1)
+		seed := seedHash(c.epoch*epochLength + 1)
+
+		if dir == "" {
+			c.cache = make([]uint32, size/4)
+			generateCache(c.cache, seed)
+			return
+		}
+		// Disk storage is needed, this will get fancy
+		endian := "le"
+		if !isLittleEndian() {
+			endian = "be"
+		}
+		path := filepath.Join(dir, fmt.Sprintf("cache-R%d-%x.%s", algorithmRevision, seed, endian))
+		logger := log.New("seed", hexutil.Bytes(seed))
+
+		// Try to load the file from disk and memory map it
+		var err error
+		c.dump, c.mmap, c.cache, err = memoryMap(path, false)
+		if err == nil {
+			logger.Info("Loaded old ethash cache from disk")
+			return
+		}
+		logger.Debug("Failed to load old ethash cache", "err", err)
+
+		// No previous cache available, create a new cache file to fill
+		if err = os.MkdirAll(filepath.Dir(path), os.ModePerm); err != nil {
+			logger.Error("Failed to create ethash cache dir", "err", err)
+		} else {
+			// Create a huge empty file to fill with the cache
+			dump, err := os.Create(path)
+			if err != nil {
+				logger.Error("Failed to create new ethash cache", "err", err)
+			} else if err = dump.Truncate(int64(size)); err != nil {
+				logger.Error("Failed to allocate new ethash cache", "err", err)
+			} else {
+				// Dump allocates, map for writing and fill
+				dump.Close()
+
+				c.dump, c.mmap, c.cache, err = memoryMap(path, true)
+				if err != nil {
+					logger.Error("Failed to memory map ethash cache", "err", err)
+				}
+			}
```

Comment:
Thanks for the test. Albeit I figured caches won't conflict since they go into different folders and I don't expect people to mine with multi processes (especially with dag rotation). The problem nonetheless stands. Will try to fix.

A side effect is that if geth crashes (terminates) while generating a DAG/cache, currently that will corrupt it and we'll never know. So this is actually a better solution.